import.py

```diff
@@ -583,6 +583,80 @@ def write_unified_cpp_file(dir):
       for cpp_file in cpps:
         f.write("#include \"{}\"\n".format(cpp_file))
 
+def extract_pref_from_line(line):
+    m = re.match(r'pref\((".*?"),\s*(.*)\);', line)
+    if not m:
+        print("extract_pref_from_line() failed on {}".format(line))
+        sys.exit(1)
+    name = m.group(1)
+    val = m.group(2)
+    kind = ""
+    if val == "true" or val == "false":
+        kind = "bool"
+    elif val.startswith('"') and val.endswith('"'):
+        kind = "string"
+    else:
+        kind = "int"
+    return (name, val, kind)
+
+def is_preprocessor_line(line):
+    return (line.startswith("#if") or
+            line.startswith("#endif") or
+            line.startswith("#else"))
+
+def extract_preferences(lines):
+    # Returns a triple of lists of ((pref,value),[include-guard]).
+    #   ([((bool_pref, value), [guards])],
+    #    [((int_pref, value), [guards])],
+    #    [((string_pref, value), [guards])])
+    i = 0
+    guards = []
+    bool_prefs = []
+    int_prefs = []
+    string_prefs = []
+    prefs = dict({"bool": [], "int": [], "string": []});
+    while i < len(lines):
+        line = lines[i]
+        if line.startswith("#if"):
+            guards.append(line)
+        elif line.startswith("pref"):
+            (pref, value, kind) = extract_pref_from_line(line)
+            prefs[kind] += [((pref, value), guards[:])]
+        elif line.startswith("#endif"):
+            guards.pop()
+        i += 1
+    # Sort the lists based on pref name, so we can binary search it in
+    # the Preferences implementation.
+    bool_prefs = list(sorted(prefs["bool"], key=lambda x: x[0][0]))
+    int_prefs = list(sorted(prefs["int"], key=lambda x: x[0][0]))
+    string_prefs = list(sorted(prefs["string"], key=lambda x: x[0][0]))
+    return (bool_prefs, int_prefs, string_prefs)
+
+def write_pref_group(dst, prefs, prefix, pref_type):
+    dst.write("static const " + pref_type + " s" + prefix + "" + pref_type + "s[] = {\n")
+    for ((name, value),guards) in prefs:
+        for guard in guards:
+            dst.write(guard)
+        dst.write("  { " + name + ", " + value + " },\n")
+        for guard in guards:
+            dst.write("#endif // " + guard)
+    dst.write("};\n\n")
+
+def copy_media_prefs(src_pref_file, dst_pref_file, prefix):
+    is_media_pref = re.compile(r'pref\("media.')
+    with open(src_pref_file, "r") as src:
+        # Filter out non media pref and non preprocessor lines.
+        lines = []
+        for line in src:
+            if is_media_pref.match(line) or is_preprocessor_line(line):
```

Comment:
This seems to contradict the comment above. Preprocessor lines seem to be added to the lines list, IIUC.

Reply:
Ok nevermind that comment, I misread the code :)